File: 05. Singleton/in Swift/ClassicalSingleton.swift
class ClassicalSingleton {
    private static var uniqueInstance: ClassicalSingleton?; // thread-safe 초기화 X
    
    private init() {}

    static func getInstance() -> ClassicalSingleton { 
        if let singleton: ClassicalSingleton = uniqueInstance {
            return singleton
        }
        
        uniqueInstance = ClassicalSingleton()
        return uniqueInstance!
    }
}

class ClassicalSingleton2 {
    // 메모리에 크게 문제가 되지 않는 경우 나중에 만들 필요는 없다.
    private static let uniqueInstance = ClassicalSingleton2() // thread-safe 초기화

    private init() {}

    static func getInstance() -> ClassicalSingleton2 {
        return uniqueInstance
    }
}

File: 05. Singleton/in JAVA/ClassicalSingleton.java
public class ClassicalSingleton {
    // Singleton 클래스의 유일한 인스턴스를 저장
    private static ClassicalSingleton uniqueInstance; 

    private ClassicalSingleton() { } // 생성자 은닉화
    
    public static ClassicalSingleton getInstance() {
        if (uniqueInstance == null) { // 첫 호출
                uniqueInstance = new ClassicalSingleton(); // 생성
        }
        return uniqueInstance;
    }
    // 기타 메소드
}
